File: mesmaeker-mockito/src/main/java/fr/xgouchet/mesmaeker/mockito/MockitoContract.java
package fr.xgouchet.mesmaeker.mockito;

import org.mockito.Mockito;

import java.util.function.Function;

import fr.xgouchet.mesmaeker.core.Contract;


/**
 * @author Xavier F. Gouchet
 */
public class MockitoContract<T> extends Contract<T> {

    public void useMockInstance(Class<T> tClass) {
        setMockInstance(Mockito.mock(tClass));
    }

    public <O> OngoingClause<T, O> whenever(Function<T, O> function) {
        if (isInstanceMock()) {
            return new MockitoOngoingClause<>(getInstance(), function);
        } else {
            return new ValidatingOngoingClause<>(getInstance(), function);
        }
    }

    public <O> OngoingClause<T, O> when(Function<T, O> function) {
        if (isInstanceMock()) {
            return new MockitoOngoingClause<>(getInstance(), function);
        } else {
            return new ValidatingOngoingClause<>(getInstance(), function);
        }
    }

    public OngoingThrowingClause<T> doThrow(Throwable... throwables) {
        if (isInstanceMock()) {
            return new MockitoOngoingThrowingClause<>(getMock(), throwables);
        } else {
            return new ValidatingOngoingThrowingClause<>(getInstance(), throwables);
        }
    }

    // Additional method helps users of JDK7+ to hide heap pollution / unchecked generics array creation
    @SuppressWarnings({"unchecked", "varargs"})
    public OngoingThrowingClause<T> doThrow(Class<? extends Throwable>... throwableClasses) {
        if (isInstanceMock()) {
            return new MockitoOngoingThrowingClause<>(getMock(), throwableClasses);
        } else {
            return new ValidatingOngoingThrowingClause<>(getInstance(), throwableClasses);
        }
    }

    // Additional method helps users of JDK7+ to hide heap pollution / unchecked generics array creation
    @SuppressWarnings({"unchecked", "varargs"})
    public <O> OngoingReturningClause<T, O> doReturn(O value, O... values) {
        if (isInstanceMock()) {
            return new MockitoOngoingReturningClause<>(getInstance(), value, values);
        } else {
            return new ValidatingOngoingReturningClause<>(getInstance(), value, values);
        }
    }

    public OngoingVoidClause<T> doNothing() {
        if (isInstanceMock()) {
            return new MockitoOngoingVoidClause<>(getMock());
        } else {
            return new ValidatingOngoingVoidClause<>(getInstance());
        }
    }


    // TODO public static <O> OngoingThrowingClause<T, O> doAnswer(Answer answer) {    }
}

File: mesmaeker-mockito/src/main/java/fr/xgouchet/mesmaeker/mockito/MockitoContract.kt
package fr.xgouchet.mesmaeker.mockito

import fr.xgouchet.mesmaeker.core.BaseContract
import org.mockito.Mockito


/**
 * TODO add javadoc !
 */
abstract class MockitoContract<T>(val mockedClass: Class<T>) : BaseContract<T>() {

    override fun instantiateMock(): T {
        return Mockito.mock(mockedClass)
    }

    fun <O> whenever(function: (T) -> O): OngoingClause<T, O> {
        return if (isInstanceMock) {
            MockitoOngoingClause(getMock(), function)
        } else {
            ValidatingOngoingClause(getInstance(), function)
        }
    }

    fun <O> `when`(function: (T) -> O): OngoingClause<T, O> {
        return if (isInstanceMock) {
            MockitoOngoingClause(getMock(), function)
        } else {
            ValidatingOngoingClause(getInstance(), function)
        }
    }

    fun doThrow(vararg throwables: Throwable): OngoingThrowingClause<T> {
        return if (isInstanceMock) {
            MockitoOngoingThrowingClause(getMock(), arrayOf(*throwables))
        } else {
            ValidatingOngoingThrowingClause(getInstance(), *throwables)
        }
    }

    // Additional method helps users of JDK7+ to hide heap pollution / unchecked generics array creation
    fun doThrow(vararg throwableClasses: Class<out Throwable>): OngoingThrowingClause<T> {
        return if (isInstanceMock) {
            MockitoOngoingThrowingClause(getMock(), *throwableClasses)
        } else {
            ValidatingOngoingThrowingClause(getInstance(), *throwableClasses)
        }
    }

    // Additional method helps users of JDK7+ to hide heap pollution / unchecked generics array creation
    fun <O> doReturn(value: O, vararg values: O): OngoingReturningClause<T, O> {
        return if (isInstanceMock) {
            MockitoOngoingReturningClause(getMock(), value, *values)
        } else {
            ValidatingOngoingReturningClause(getInstance(), value, *values)
        }
    }

    fun doNothing(): OngoingVoidClause<T> {
        return if (isInstanceMock) {
            MockitoOngoingVoidClause(getMock())
        } else {
            ValidatingOngoingVoidClause(getInstance())
        }
    }


    // TODO public static <O> OngoingThrowingClause<T, O> doAnswer(Answer answer) {    }
}
